HTML API: Add `matches_breadcrumbs()` to HTML Processor for better querying

dmsnell opened the pull request:
Trac ticket: [#59400](https://core.trac.wordpress.org/ticket/59400#ticket)

 - [x] Cleared for Gutenberg WPCS issues in WordPress/gutenberg#54629

Inside a `next_tag()` loop it can be challenging to use breadcrumbs because they are only exposed inside the call to `next_tag()` via the `$query` arg.

In this patch a new method, `matches_breadcrumbs()` is exposed which allows for querying within the `next_tag()` loop for more complicated queries.

This method exposes a wildcard `*` operator to allow matching _any HTML tag_ that the currently-matched tag is a child or descendant of.

dmsnell committed:
HTML API: Add `matches_breadcrumbs()` to HTML Processor for better querying

Inside a `next_tag()` loop it can be challenging to use breadcrumbs because
they are only exposed inside the call to `next_tag()` via the `$query` arg.

In this patch a new method, `matches_breadcrumbs()` is exposed which allows
for querying within the `next_tag()` loop for more complicated queries.

This method exposes a wildcard `*` operator to allow matching _any HTML tag_
that the currently-matched tag is a child or descendant of.

mukeshpanchal27 commented on tests/phpunit/tests/html-api/wpHtmlProcessorBreadcrumbs.php:
```diff
@@ -352,6 +352,45 @@ public function data_html_target_with_breadcrumbs() {
+			'Inner IMG'                  => array( '<div><span><figure><img target></figure></span></div>', array( 'span', 'figure', 'img' ), true ),
+			'Inner IMG wildcard'         => array( '<div><span><figure><img target></figure></span></div>', array( 'span', '*', 'img' ), true ),
+			'Inner IMG no wildcard'      => array( '<div><span><figure><img target></figure></span></div>', array( 'span', 'img' ), false ),
+			'Full specification'         => array( '<div><span><figure><img target></figure></span></div>', array( 'html', 'body', 'div', 'span', 'figure', 'img'), true ),
+			'Invalid Full specification' => array( '<div><span><figure><img target></figure></span></div>', array( 'html', 'div', 'span', 'figure', 'img'), false ),
```
```suggestion
			'Full specification'         => array( '<div><span><figure><img target></figure></span></div>', array( 'html', 'body', 'div', 'span', 'figure', 'img' ), true ),
			'Invalid Full specification' => array( '<div><span><figure><img target></figure></span></div>', array( 'html', 'div', 'span', 'figure', 'img' ), false ),
```

Fix PHPCS failure

dmsnell replied:
Thanks @mukeshpanchal27 - I've incorporated your change into ab6952c216

Please feel free to overlook code-styling issues for now, both because this isn't ready for that stage of review, and also because the CI job should point all these out, and I wouldn't want you to spend your time doing what the automated system will do.

dmsnell committed:
Replace `next_tag()` logic with new method

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -406,26 +406,66 @@ public function next_tag( $query = null ) {
+	 *
+	 * A "*" represents a single tag wildcard, where any tag matches, but not no tags.
+	 *
+	 * At some point this function _may_ support a `**` syntax for matching any number
+	 * of unspecified tags in the breadcrumb stack. This has been intentionally left
+	 * out, however, to keep this function simple and to avoid introducing backtracking,
+	 * which could open up surprising performance breakdowns.
```
So let's say you did want to implement `**` behavior but limited to 5 levels, where `img` is, say, inside of `main`. To do that, you currently would have to do:

```php
for ( $asterisk_count = 0; $asterisk_count <= 5; $asterisk_count++ ) {
    if (
        $processor->matches_breadcrumbs(
            array(
                'main',
                ...array_fill( 0, $asterisk_count, '*' )
                'img' 
            ) 
        ) 
    ) {
        return true;
    }
}
return false;
```

Is that right?

dmsnell replied:
ha yeah this would work and accurately conveys the backtracking problem, though I believe that there are some more performant backtracking mechanisms that don't have to re-process as much as this.

dmsnell replied:
it's still possible that I'm way overthinking this, confusing it with a different kind of match, where it would matter which of the inner or outer elements are matching.

for a pattern like `main, **, div, **, img` though I want to make sure this doesn't break, or maybe it's better to compare that against `main, *, div, **, img`

for the first one, with two `**` groups, I read it as "walk up the stack of open elements; if there is a DIV somewhere and also a MAIN somewhere then this matches."

for the second one though we have a tighter constraint above the simpler one. if we have this breadcrumb `HTML > BODY > MAIN > SECTION > DIV > DIV > IMG` then we have to make sure that the `**` group _only_ matches the last DIV, otherwise `main, *, div` will fail to match because we were too greedy.

This is where I think we must start backtracking: after that higher DIV fails the rest of the match. That's a very recursive problem too. If we start by being un-greedy though, we stop at the lower DIV and it fails the rest of the match, so we have to track that `**` group and backtrack on the pattern vs. on the content. We have to remember each double wildcard group and go back to reattempt matches after adjusting our position in the stack.

The probably probably starts if `*` allowed for a zero match, but when I tossed it in there it didn't seem to have any connotation of that since this is like a filesystem wildcard where those expectations are that `*` in a path is _something_ and not _nothing_.

dmsnell replied:
thinking all this through pushes me to leave this for userspace for now. we expose `get_breadcrumbs()` and so someone is implement this on their own until we find a good solution for it.

```php
function preg_matches_breadcrumbs( string $pattern, array $breadcrumbs ) {
	return false !== preg_match( $pattern, implode( ' > ', $breadcrumbs ) );
}

if ( preg_matches_breadcrumbs( '~ > MAIN > .* IMG$~', $processor->get_breadcrumbs() ) ) {
	…
}
```

this format is surprisingly useful now that I hack it up

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -406,26 +406,66 @@ public function next_tag( $query = null ) {
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string[] $breadcrumbs list of HTML tag names representing nested structure.
```
```suggestion
	 * @param string[] $breadcrumbs list of HTML tag names representing nested structure. A wildcard tag name '*' is also allowed.
```

dmsnell replied:
thanks. I should have earmarked this and any other places where breadcrumbs are being normalized, including in `next_tag()`

dmsnell replied:
Added in both docblocks 👍

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -406,26 +406,66 @@ public function next_tag( $query = null ) {
+		}
+
+		// Start at the last crumb.
+		$crumb_at = count( $breadcrumbs ) - 1;
```
While not likely, it's possible that `$breadcrumbs` may not have indicies that increment uniformly. For example, if someone had these breadcrumbs:

```php
$breadcrumbs = array(
    'html',
    'body',
    'main',
    'img',
);
if ( ! $use_main ) {
    $breadcrumbs = array_diff( $breadcrumbs, array( 'main' ) );
}
```

Then `$breadcrumbs` will now be:

```php
array (
  0 => 'html',
  1 => 'body',
  3 => 'img',
)
```

This would break the iteration below.

So should perhaps this be made more robust by passing `$breadcrumbs` through `array_values()` first?

```suggestion
		$breadcrumbs = array_values( $breadcrumbs );
		$crumb_at = count( $breadcrumbs ) - 1;
```

dmsnell replied:
interesting. at first I used `end()`/`current()`/`prev()` but changed to the index to avoid mutating the array. I guess I could go back to those.

I'm nervous to start with code that causes unnecessary memory allocation or uncertain runtime performance, even if they are for expectedly small arrays. any thoughts?

dmsnell replied:
silly me. we're copying these arrays by passing them into the functions, so I guess we won't risk mutating the passed arrays, and we won't save anything by trying to avoid allocation.

I've restored `end()` and `current()` and `prev()` because of this. I don't think that passing arrays by reference clearly saves on allocations either, but I'm curious to explore that route.

dmsnell committed:
PR feedback

 - add wildcard to docblock description of breadcrumbs parameters
 - revert to scanning breadcrumb array with end/current/prev to handle sparse arrays

Props: @westonruter

dmsnell committed:
Ensure that wildcards match the current tag as well

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -406,26 +407,67 @@ public function next_tag( $query = null ) {
+
+		if ( '*' !== $crumb && $this->get_tag() !== strtoupper( $crumb ) ) {
+			return false;
+		}
```
this is bothering me because I don't currently understand why it's necessary. it seems like we should have the currently-matched tag on the stack of open elements, but we still need to check this. then, it seems like we'd be checking twice (in the first iteration of the loop below), but we aren't and all the tests are passing.

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -406,26 +407,67 @@ public function next_tag( $query = null ) {
+	 *                              May also contain the wildcard `*` which matches a single element, e.g. `array( 'SECTION', '*' )`.
+	 * @return bool Whether the currently-matched tag is found at the given nested structure.
+	 */
+	public function matches_breadcrumbs( $breadcrumbs ) {
```
Maybe this is out of scope for this PR, but could the `$breadcrumbs` format be expanded beyond just a list of tag names to instead be a list of queries (or the like)? That is, it would be very useful to check not just that the tag name matches but more so that the element has a given ID or a class name. Otherwise, the issue is that tag name alone may not be specific enough to be useful.

westonruter replied:
For example, let's say I have the HTML for the entire page and I need to add `fetchpriority=high` to the featured image. In Chrome DevTools, the featured image has the selector `#page > div.single-featured-image-header > img`. In tag name breadcrumbs, this is `['div', 'div', 'img']` which is very unspecific. It would be great to be able to pass in something like:

```php
array(
	array("id" => "page"),
	array(
		"tag" => "div",
		"class" => "single-featured-image-header",
	),
	array("tag" => "img"),
}
```

dmsnell replied:
it's definitely out of scope but it's something I've been thinking about a lot recently. it extends to all sorts of challenges in querying and I've been exploring the use of bookmarks to accomplish these needs. one thing we probably don't want to do here is jump around in the document _while jumping around_, meaning either we track everything about every token (memory bloat) or find a way to only trap what we care about when.

the HTML Processor is using `$current_token` to hold not only a tag name, but also a full reference to the location in the document for where something is. this is actually required to properly run the parsing, as certain operations imply closing tags of not only a given name, but a specific tag.

the idea would be something like this, where we still leave this up to userspace so that the performance of the processor remains more predictable, and additional processing/memory can be exposed explicitly in consuming code.

```php
while ( $processor->next_tag() ) {
	if (
		'DIV' === $processor->get_tag() &&
		$processor->has_class( 'single-featured-image-header' )
	) {
		$processor->set_bookmark( 'wrapper' );
		continue;
	}

	if (
		'IMG' === $processor->get_tag() &&
		$processor->is_bookmark_open( 'wrapper' )
	) {
		$processor->set_attribute( 'fetchpriority', 'high' );
	}
```

and I believe that we could create a variety of relationship functions like this: `is_child_of( $bookmark_name )`, `is_descendant_of( $bookmark_name )`, `is_sibling_of( $bookmark_name )`.

internally there's a callback for when popping an element off of the stack of open elements, and I have been thinking of exposing that to userspace. I _think_ it should be safe, but I have questions about how best to represent that. this is what led me to the `is_` functions with bookmarks.

ockham commented:
Rebased.

ockham commented:
Committed to Core in https://core.trac.wordpress.org/changeset/56702.
